Conjugate gradient

joanrue opened the pull request:
Added the Conjugate Gradient Method iterative solver based on the Sepand's Solver.py base class.

joanrue committed:
[solver] add conjugate gradient method

Added the Conjugate Gradient Method iterative solver based on the Sepand;;s Solver.py base class

joanrue committed:
[abc/solver] add base class initialisation in _Worker

_Worker was not initializing base class threading.Thread, this was raising the error 'Thread.__init__() was not called'

joanrue committed:
[solver/cg] Change np.einsum to module agnostic funcs (__mul__ & __add__)

joanrue committed:
[solver/cg] Change numpy code to module agnostic based on input data

joanrue committed:
[solver/cg] Fix module agnosticity and precision enforcement.

Tested with numpy cupy and dask arrays

joanrue committed:
[solver/CG] Black formatting

SepandKashani commented on src/pycsou/abc/solver.py:
```diff
@@ -566,6 +566,7 @@ def _m_persist(self):
 
     class _Worker(threading.Thread):
         def __init__(self, solver: "Solver"):
+            super().__init__()
```
ok based on our discussion.

SepandKashani commented on src/pycsou/opt/solver/cg.py:
```diff
@@ -0,0 +1,142 @@
+       {\mathbf{A}\mathbf{x} = \mathbf{b}},
+    for the vector :math:`\mathcal{x}: \mathbb{x}^N`.
+    where:
+    * :math:`\mathcal{A}: \mathbb{R}^M\times^N` is a *symmetric* *positive definite* matrix.
```
*positive def* OPERATOR, not necessarily a matrix. (We only need to know how to apply the op afterall.)

SepandKashani commented on src/pycsou/opt/solver/cg.py:
```diff
@@ -0,0 +1,142 @@
+    def fit(
+        self,
+        x0: pyct.NDArray = None,
+        stop_crit: pycs.StoppingCriterion = None,
```
If `tol` is provided in `fit()`, then you should not have `stop_crit` in the signature anymore.
It is the job of `fit()` to construct the AbsErr stop_crit and call `super().fit()` with it.

SepandKashani commented on src/pycsou/opt/solver/cg.py:
```diff
@@ -0,0 +1,142 @@
+        r"""
+        Solve the minimization problem defined in :py:meth:`CG.__init__`, with the provided
+        run-specific parameters.
+        Parameters
```
Add an empty line before `Parameters`.

SepandKashani commented on src/pycsou/opt/solver/cg.py:
```diff
@@ -0,0 +1,142 @@
+        Parameters
+        ----------
+        x0: NDArray
+           (..., N) Starting guess(es) for the solution(s). Defaults to a zero NDArray if unspecified.
```
Prefer the term "initial point" to "starting guess".

SepandKashani commented on src/pycsou/opt/solver/cg.py:
```diff
@@ -0,0 +1,142 @@
+        p = mst["p"]
+
+        ap = self._a.apply(p)
+        rr = (r * r).sum(-1)[..., None]
```
`rr = xp.linalg.norm(r, ord=2, axis=-1, keepdims=True` should work and avoids 2 temporary arrays in the process.

SepandKashani commented on src/pycsou/opt/solver/cg.py:
```diff
@@ -0,0 +1,142 @@
+
+        ap = self._a.apply(p)
+        rr = (r * r).sum(-1)[..., None]
+        alpha = rr / (p * ap).sum(-1)[..., None]
```
`rr / (p*ap).sum(axis=-1, keepdims=True)` avoids having to do the `[...,None]` indexing at the end.

SepandKashani commented on src/pycsou/opt/solver/cg.py:
```diff
@@ -0,0 +1,142 @@
+        alpha = rr / (p * ap).sum(-1)[..., None]
+        x = x + alpha * p
+        r = r - alpha * ap
+        beta = (r * r).sum(-1)[..., None] / rr
```
Same comment as above.

joanrue committed:
[solver/CG] add Sepand's comments

use xp.norm instead of sum(r * r) and remove the stop_crit input

SepandKashani commented on src/pycsou/opt/solver/cg.py:
```diff
@@ -0,0 +1,132 @@
+       {\mathbf{A}\mathbf{x} = \mathbf{b}},
+    for the vector :math:`\mathcal{x}: \mathbb{x}^N`.
+    where:
+    * :math:`\mathcal{A}: \mathbb{R}^M\times^N` is a *symmetric* *positive definite* operator.
```
You use both M and N as indices here, but A must be square.
Double-check the whole docstring and standardize on one letter only.

Also, rewrite this as `\mathcal{A}: \mathbb{R}^{N} \to \mathbb{R}^{N}`

SepandKashani replied:
In `CG.fit()`, the docstring says "the minimization problem defined in :py:meth:`CG.__init__`", but Ax=b is not a minimization problem.
You should instead write \argmin_{x\in\bR^{N}} ||Ax - b||_{2}^{2}, where A is ...

SepandKashani commented on src/pycsou/opt/solver/cg.py:
```diff
@@ -0,0 +1,132 @@
+        self,
+        a: pyco.PosDefOp,
+        b: pyct.NDArray,
+        log_var: pyct.VarName = ("x",),
```
I missed this last time: why are `folder`, `exist_ok`, `writeback_rate`, `verbosity` missing here?

SepandKashani commented on src/pycsou/opt/solver/cg.py:
```diff
@@ -0,0 +1,132 @@
+
+    def m_init(self, x0: pyct.NDArray, tol: float = 1e-5):
+        self._mstate["x"] = x0
+        self._mstate["r"] = r = self._b - self._a.apply(x0)
```
`= r =` is useless.

SepandKashani commented on src/pycsou/opt/solver/cg.py:
```diff
@@ -0,0 +1,132 @@
+        p = r + beta * p
+        mst["x"], mst["r"], mst["p"] = x, r, p
+
+    @pycrt.enforce_precision(o=True)
```
No decorator should be used here. If your `m_init` and `m_step` are correct, the output dtype should match `b` and `x0`s dtype.

SepandKashani commented on src/pycsou/opt/solver/cg.py:
```diff
@@ -0,0 +1,132 @@
+        xp = pycu.get_array_module(self._b)
+        x0 = x0 if (x0 is None) else pycrt.coerce(x0)
+
+        try:
```
`AbsError` will check `tol > 0`: don't need to test again here.

SepandKashani commented on src/pycsou/opt/solver/cg.py:
```diff
@@ -0,0 +1,132 @@
+                    f"(array module {pycu.get_array_module(x0)} is different from {xp}."
+                )
+        else:
+            x0 = xp.zeros((1, self._b.shape[-1]), dtype=pycrt.getPrecision().value)
```
`dtype = self._b.dtype`.

SepandKashani commented on src/pycsou/opt/solver/cg.py:
```diff
@@ -0,0 +1,132 @@
+import pycsou.util.ptype as pyct
+
+
+def is_broadcastable(arr1, arr2):
```
Use `numpy.broadcast_shapes()` instead of rolling out your own function.

SepandKashani commented on src/pycsou/opt/solver/cg.py:
```diff
@@ -0,0 +1,132 @@
+        else:
+            x0 = xp.zeros((1, self._b.shape[-1]), dtype=pycrt.getPrecision().value)
+
+        return super().fit(x0, mode=mode, stop_crit=stop_crit)
```
I now have a doubt on this. Leave as-is and see if it runs.

SepandKashani reviewed (changes requested):
Looking better. Still some things to tackle.

joanrue committed:
[solver/CG] add Sepand's comments

change docstring of cg.fit(), use numpy broadcast test instead of own, add base class inputs to cg.init, remove enforce_precision decoration of cg.solution()
